tiledb/cloud/client.py

```diff
@@ -375,148 +372,154 @@ def find_organization_or_user_for_default_charges(user):
     return namespace_to_charge
 
 
+class RetryMode(enum.Enum):
+    DEFAULT = "default"
+    FORCEFUL = "forceful"
+    DISABLED = "disabled"
+
+    def maybe_from(v: "RetryOrStr") -> "RetryMode":
+        if isinstance(v, RetryMode):
+            return v
+        return RetryMode(v)
+
+
+RetryOrStr = Union[RetryMode, str]
+
+
+_RETRY_CONFIGS = {
+    RetryMode.DEFAULT: urllib3.Retry(
+        total=10,
+        backoff_factor=0.25,
+        status_forcelist=[503],
+        allowed_methods=[
+            "HEAD",
+            "GET",
+            "PUT",
+            "DELETE",
+            "OPTIONS",
+            "TRACE",
+            "POST",
+            "PATCH",
+        ],
+        raise_on_status=False,
+        # Don't remove any headers on redirect
+        remove_headers_on_redirect=[],
+    ),
+    RetryMode.FORCEFUL: urllib3.Retry(
+        total=10,
+        backoff_factor=0.25,
+        status_forcelist=[400, 500, 501, 502, 503],
+        allowed_methods=[
+            "HEAD",
+            "GET",
+            "PUT",
+            "DELETE",
+            "OPTIONS",
+            "TRACE",
+            "POST",
+            "PATCH",
+        ],
+        raise_on_status=False,
+        # Don't remove any headers on redirect
+        remove_headers_on_redirect=[],
+    ),
+    RetryMode.DISABLED: False,
+}
+
+
+# Type of the callback function we pass the response UUID.
+IDCallback = Callable[[Optional[uuid.UUID]], Any]
+
+
+def send_udf_call(
+    api_func: Callable[..., urllib3.HTTPResponse],
+    api_kwargs: Dict[str, Any],
+    decoder: results.AbstractDecoder,
+    *,
+    id_callback: IDCallback,
+    store_results: bool,
+) -> results.Response:
+    """Sends an API request and returns the results in the form of a Response."""
+    try:
+        http_response = api_func(_preload_content=False, **api_kwargs)
+    except rest_api.ApiException as exc:
+        id_callback(results.extract_task_id(exc))
+        raise tiledb_cloud_error.check_exc(exc) from exc
+
+    task_id = results.extract_task_id(http_response)
```

Comment:
Do we want to catch here? This function will throw if the future throws -- if I'm following correctly, when you get the result then any exception on the "thread" will propagate.

Reply:
This is the synchronous version, and `extract_task_id` is guaranteed never to throw anything; it will either return None or a UUID.

I’ve expanded the documentation on this function to make it a bit clearer what is going on.

Reply:
Won't this line throw if there's an exception on the thread? https://github.com/TileDB-Inc/TileDB-Cloud-Py/pull/181/files#diff-ab11ae7abf2b52119b99ba2775df89d3baeaf5e7062dc8cb131d911bc9033c5bR118

x-ref: https://docs.python.org/3/library/concurrent.futures.html#concurrent.futures.Future.result
> If the call raised an exception, this method will raise the same exception.

Reply:
This *is* the thread; when we make a call without `_async_req=True` (and with `_preload_content=False`), the generated API returns an immediate `urllib3.HTTPResponse`, not a `futures.Future[urllib3.HTTPResponse]`. (Out of all the things I hate about the generated code, the fact that providing different parameters *wildly* changes the type it returns, and in *incredibly* poorly-documented ways, is probably the worst.)

On the line you’re referencing, both `HTTPResponse` and `ApiException` are guaranteed to have a `.headers` field which is (or at least can be treated as) an `Optional[Mapping[str, str]]`.

Reply:
I see, thanks.